src/com/t_oster/liblasercut/drivers/K40NanoDriver.java

```diff
@@ -0,0 +1,1484 @@
+  {
+    return this.bedHeight;
+  }
+
+  /**
+   * This method should return a name for this driver.
+   *
+   * @return
+   */
+  @Override
+  public String getModelName()
+  {
+    return "K40 Stock-LIHUIYU M2/M1/M/B2/B1/B/A";
+  }
+
+  /**
+   * This method must copy the current instance with all config settings,
+   * because it is used for save- and restoring
+   *
+   * @return
+   */
+  @Override
+  public LaserCutter clone()
+  {
+    K40NanoDriver clone = new K40NanoDriver();
+    clone.bedHeight = this.bedHeight;
+    clone.bedWidth = this.bedWidth;
+    clone.board = this.board;
+    clone.mock = this.mock;
+    return clone;
+  }
+
+  private static final String[] settingAttributes = new String[]
+  {
+    SETTING_BEDWIDTH, SETTING_BEDHEIGHT, SETTING_BOARD, SETTING_MOCK
+  };
+
+  @Override
+  public String[] getPropertyKeys()
+  {
+    return settingAttributes;
+  }
+
+  @Override
+  public void setProperty(String attribute, Object value)
+  {
+    if (SETTING_BEDWIDTH.equals(attribute))
+    {
+      this.setBedWidth((Double) value);
+    }
+    else if (SETTING_BEDHEIGHT.equals(attribute))
+    {
+      this.setBedHeight((Double) value);
+    }
+    else if (SETTING_BOARD.equals(attribute))
+    {
+      this.setBoard((String) value);
+    }
+    else if (SETTING_MOCK.equals(attribute))
+    {
+      this.setMock((Boolean) value);
+    }
+  }
+
+  @Override
+  public Object getProperty(String attribute)
+  {
+    if (SETTING_BEDWIDTH.equals(attribute))
+    {
+      return this.getBedWidth();
+    }
+    else if (SETTING_BEDHEIGHT.equals(attribute))
+    {
+      return this.getBedHeight();
+    }
+    else if (SETTING_BOARD.equals(attribute))
+    {
+      return this.getBoard();
+    }
+    else if (SETTING_MOCK.equals(attribute))
+    {
+      return this.isMock();
+    }
+    return null;
+  }
+
+  public class K40Device
+  {
+
+    private static final int UNINIT = 0;
+    private static final int DEFAULT = 1;
+    private static final int COMPACT = 2;
+
+    static final char LASER_ON = 'D';
+    static final char LASER_OFF = 'U';
+
+    static final char RIGHT = 'B';
+    static final char LEFT = 'T';
+    static final char TOP = 'L';
+    static final char BOTTOM = 'R';
+    static final char DIAGONAL = 'M';
+
+    K40Queue jobber;
+    private StringBuilder builder = new StringBuilder();
+
+    private int mode = UNINIT;
+
+    private boolean is_top = false;
+    private boolean is_left = false;
+    private boolean is_on = false;
+
+    private double speed = 30;
+    private double power = 1;
+    private String board = "M2";
+
+    private int x = 0;
+    private int y = 0;
+
+    void open()
+    {
+      jobber = new K40Queue();
+      jobber.open();
+    }
+
+    void close()
+    {
+      if (mode == COMPACT)
+      {
+        exit_compact_mode();
+        execute();
+      }
+      jobber.close();
+      jobber = null;
+    }
+
+    void setPower(double power)
+    {
+      this.power = power;
```

Comment:
it should be documented that setPower does nothing because the device doesn't support it